src/vm/amd64/PInvokeStubs.asm

```diff
@@ -135,4 +143,91 @@ NESTED_ENTRY VarargPInvokeGenILStub, _TEXT
 
 NESTED_END VarargPInvokeGenILStub, _TEXT
 
+;
+; in:
+; InlinedCallFrame (rcx) = pointer to the InlinedCallFrame data, including the GS cookie slot (GS cookie right 
+;                          before actual InlinedCallFrame data)
+;
+;
+LEAF_ENTRY JIT_PInvokeBegin, _TEXT
+        
+        mov             rax, qword ptr [s_gsCookie]
+        mov             qword ptr [rcx], rax
+        add             rcx, SIZEOF_GSCookie
+
+        ;; set first slot to the value of InlinedCallFrame::`vftable' (checked by runtime code)
+        lea             rax,[??_7InlinedCallFrame@@6B@]
+        mov             qword ptr [rcx], rax
+
+        mov             qword ptr [rcx + OFFSETOF__InlinedCallFrame__m_Datum], 0
+
+        mov             rax, rsp
+        add             rax, 8
+        mov             qword ptr [rcx + OFFSETOF__InlinedCallFrame__m_pCallSiteSP], rax
+        mov             qword ptr [rcx + OFFSETOF__InlinedCallFrame__m_pCalleeSavedFP], rbp
+
+        mov             rax, [rsp]
+        mov             qword ptr [rcx + OFFSETOF__InlinedCallFrame__m_pCallerReturnAddress], rax
+
+        INLINE_GETTHREAD rax 
+        ;; pFrame->m_Next = pThread->m_pFrame;
+        mov             rdx, qword ptr [rax + OFFSETOF__Thread__m_pFrame]
+        mov             qword ptr [rcx + OFFSETOF__Frame__m_Next], rdx
+
+        ;; pThread->m_pFrame = pFrame;
+        mov             qword ptr [rax + OFFSETOF__Thread__m_pFrame], rcx
+
+        ;; pThread->m_fPreemptiveGCDisabled = 0
+        mov             dword ptr [rax + OFFSETOF__Thread__m_fPreemptiveGCDisabled], 0
+
+        ret
+
+LEAF_END JIT_PInvokeBegin, _TEXT
+
+;
+; in:
+; InlinedCallFrame (rcx) = pointer to the InlinedCallFrame data, including the GS cookie slot (GS cookie right 
+;                          before actual InlinedCallFrame data)
+;
+;
+NESTED_ENTRY JIT_PInvokeEnd, _TEXT
+
+        alloc_stack         MIN_SIZE
+        save_reg_postrsp    r14, MIN_SIZE + 08h
```

Comment:
Can we use scratch registers like r8 or r9 here instead of using these callee saved registers?

Reply:
I need to preserve the value of the `InlineCallFrame` pointer because I need it after the call to `RareDisablePreemptiveGCHelper`. I also need the Thread pointer, but that one we could just fetch again using the INLINE_GETTHREAD macro, but i think that would be slower.
I'm using callee saved registers because this guarantees that `RareDisablePreemptiveGCHelper` will restore the values I need when it returns

Reply:
You can save the registers around the call to RareDisablePreemptiveGCHelper only if you would like to make these helpers faster.

Reply:
Good point. I refactored this and save the registers for the rare path only.